feat(rxForm): rxForm 2.0

CITguy opened the pull request:
* add `rxNestedElement` service to aid in DOM hierarchy validation
  on new directives
* lots of JSDOC improvements
  * proper "module" naming provides better search capabilities in
    generated ngdocs and easier linking between documentation
* added missing unit tests for
  * rxFieldName
  * rxCheckbox
  * rxRadio
  * rxSelect
* added new directives
  * rx-form
  * rx-form-section
  * rx-field
  * rx-field-content
  * rx-input
  * rx-prefix
  * rx-suffix
  * rx-inline-error
  * rx-help-text
* `rxForm.less` and `vars.less` updated for new code
* eliminated `rx-form-item` from demos and templates
  * `Layout 3: Create Form` markup
  * `rxFeedback` template markup
  * `rxIdentity` demo markup
  * `rxModalAction` demo markup
  * `typeahead` demo markup
* rxNotify demo updated
  * now using latest directives and rxForm patterns
  * using new layout style functionality
* `rxFieldName` moved within `rxForm`

------

* [FRMW-1](https://jira.rax.io/browse/FRMW-1)
* [FRMW-2](https://jira.rax.io/browse/FRMW-2)
* [FRMW-3](https://jira.rax.io/browse/FRMW-3)

CITguy commented:
Screenshots to follow

CITguy commented:
### rxForm
* compressed layout is no longer a thing
* rxFormFieldset is no longer a thing
  * Use headings to properly label sections
* Supports inline or stacked forms
  * Stacked via `<rx-form-section stacked>`
  * Can use inline and stacked within the same form, if desired.
* Many New Directives
  * rx-form
  * rx-form-section
  * rx-field
  * rx-field-content
  * rx-input
  * rx-prefix
  * rx-suffix
  * rx-inline-error
  * rx-help-text
* most new directives now subject to proper DOM hierarchy nesting
  * made possible via `rxMisc > rxNestedElement` service
* Updated rx-field-name to adhere to DOM hierarchy validation.
* Lots of updated documentation
  * rxForm demo
    * Updated existing examples to use new rxForm patterns.
    * In addition, new examples demonstrate some more advanced inputs.
  * Forms Styleguide

#### BEFORE
![rxform-before](https://cloud.githubusercontent.com/assets/545605/8289069/c02b2916-18df-11e5-8a02-7e187818ade3.png)

#### AFTER
![rxform-after](https://cloud.githubusercontent.com/assets/545605/8289075/c600242c-18df-11e5-83eb-a4ef964b17f1.png)

CITguy commented:
### Layout 3: Create Form
* Styleguide updated to use new rxForm patterns.

#### BEFORE
![layout3-before](https://cloud.githubusercontent.com/assets/545605/8289088/e3a5c3b0-18df-11e5-9d78-01a510419c30.png)

#### AFTER
![layout3-after](https://cloud.githubusercontent.com/assets/545605/8289090/e9ae7234-18df-11e5-807b-5d1c67b3920d.png)

CITguy commented:
### rxFeedback
* Template updated to use new rxForm patterns.

#### BEFORE
![rxfeedback-before](https://cloud.githubusercontent.com/assets/545605/8289104/fd939536-18df-11e5-810c-24ec6682b00d.png)

#### AFTER
![rxfeedback-after](https://cloud.githubusercontent.com/assets/545605/8289108/01fa719e-18e0-11e5-88d6-19ec9df9f517.png)

CITguy commented:
### rxIdentity
* Demo updated to use new rxForm patterns.

#### BEFORE
![rxidentity-before](https://cloud.githubusercontent.com/assets/545605/8289126/17fb07d8-18e0-11e5-9f0f-c7c728344b6c.png)

#### AFTER
![rxidentity-after](https://cloud.githubusercontent.com/assets/545605/8289131/1c3593ae-18e0-11e5-838c-11bf37094da4.png)

CITguy commented:
### rxModalAction
* Demo updated to use new rxForm patterns.

#### BEFORE
![rxmodalaction-password-before](https://cloud.githubusercontent.com/assets/545605/8289138/29b1f400-18e0-11e5-89cf-2e6cb54c5cb9.png)

#### AFTER
![rxmodalaction-password-after](https://cloud.githubusercontent.com/assets/545605/8289144/2fccf43e-18e0-11e5-85c9-3d611e86263a.png)

CITguy commented:
### rxNotify
* Demo updated to make use of new rxForm patterns and new directives.
* Demo also uses new `layout` functionality to place form and example side-by-side.

#### BEFORE
![rxnotify-before](https://cloud.githubusercontent.com/assets/545605/8289158/4113fbfc-18e0-11e5-8db2-aced155fb333.png)

#### AFTER
![rxnotify-after](https://cloud.githubusercontent.com/assets/545605/8289167/4707a9b4-18e0-11e5-85b7-51290c693da3.png)

CITguy commented:
### typeahead
* Demo updated to use new rxForm patterns.

#### BEFORE
![typeahead-before](https://cloud.githubusercontent.com/assets/545605/8289174/56849488-18e0-11e5-85ee-93389ec4e3d2.png)

#### AFTER
![typeahead-after](https://cloud.githubusercontent.com/assets/545605/8289179/5aeb898c-18e0-11e5-9a92-3256e082995a.png)

CITguy commented:
## rxForm Responsiveness
* rxForm styles are automatically responsive to screen width (and width of containers)
* The immediate children of `rx-form-section` (`rx-field` and `div`) will grow starting at 250px (current field width) to a max of full width. If the width cannot hold all elements, they will wrap to the next line.

**NOTE:** All three examples below are using the exact same markup.

### Wide
The `rx-field` elements will share 1/3 of the width (if wide enough)
![rxform-responsive-textinputs-large](https://cloud.githubusercontent.com/assets/545605/8289192/7531aa60-18e0-11e5-8548-cd73b966558d.png)

### Medium
The first two `rx-field` elements share 1/2 width, while the third consumes full width.
![rxform-responsive-textinputs-medium](https://cloud.githubusercontent.com/assets/545605/8289264/f2716d1c-18e0-11e5-999d-21804ca34ab4.png)

### Skinny
Same markup, `rx-field` elements wrap automatically.
![rxform-responsive-textinputs-small](https://cloud.githubusercontent.com/assets/545605/8289197/7a518ff6-18e0-11e5-9762-ba6de32e18ed.png)

CITguy commented:
FYI: mixins for vendor-prefixed CSS properties were not used as per #1024

CITguy commented:
* Addresses #948 in Styleguide > Forms
* Addresses #884

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+        <li>
+          You can use a stacked layout with these children by placing the
+          <code>stacked</code> HTML attribute on the <code>rx-form-section</code> to
+          position these children in a columnar fashion (top to bottom).
```
When reading this, I started wondering why there were html tags in a readme.  I haven't been part of the conversations around improving documentation, but I don't think it's a good idea to be putting implementation details in the styleguide.  That just means any updates have to change both places.  I would think that the styleguide just says that you can use inline or stacked fields.  It might also say *why*, but it doesn't say *how*.

CITguy replied:
Good suggestion. I'll make a tweak.

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+        <li>Please use the <code>&lt;label for="formControlId"&gt;</code> format when defining your HTML labels.
+          <ul>
+            <li>When a <code>label</code> is clicked (wherever it is placed in the DOM), it should focus or activate its corresponding form control.</li>
+            <li>If used inline with the form control, place the <code>label</code> <strong>immediately after the form control</strong>.
```
Does this apply to `<input type="text">` or `<textarea>`?

CITguy replied:
Inline? No.  I'm referring to inline with like radios and checkboxes.

megalord replied:
Ok, so are inline labels for text inputs no longer acceptable from a design standard?  If they are, I think it should be mentioned as an exception here.

CITguy replied:
Inline HTML labels should only be used for radios, checkboxes, and toggle switches because those controls are the only ones that make sense to use them.
![screen shot 2015-06-24 at 1 14 31 pm](https://cloud.githubusercontent.com/assets/545605/8337833/f68eea3a-1a72-11e5-9a3e-c0130c864a80.png)

Forms are inline when `rx-field` elements are side-by-side. Elements *within* `rx-field` will still be stacked, and HTML labels for these controls should go in the `rx-field-name`.

![screen shot 2015-06-24 at 1 13 50 pm](https://cloud.githubusercontent.com/assets/545605/8337816/e175b76e-1a72-11e5-8afb-a9f83b86e44d.png)

This will be reflected in my update.

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+                  is disabled in an <code>rxForm</code>.
+                </li>
+                <li>
+                  For consistency, and future compatibility, assume that rxTogleSwitch + label
```
~~rxTogleSwitch~~ rxToggleSwitch

CITguy replied:
good catch.

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+        <li>Do not confuse <code>rxFieldName</code> with a <code>label</code> element.
+          <ul>
+            <li><code>rxFieldName</code> is a descriptive element for a particular form field.</li>
+            <li><code>rxFieldName</code> may contain one or more <code>label</code> elements.</li>
```
Can you elaborate as to when this would be the case?

CITguy replied:
Not really, and looking at it again, the sentence doesn't make much sense. It should just say that it may contain an optional `label` element, not "one or more".

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+            </li>
+          </ul>
+        </li>
+        <li>If you need to use a button in your field for auxillary purposes, use the default blue.</li>
```
~~auxillary~~ auxiliary

CITguy replied:
good catch, will fix.

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+          disabled.
+        </li>
+        <li>
+          If you intend to use a <code>label</code> element inline, place the
```
The word "inline" throws me off here.  It makes me think that I can use a label element, but not inline, and then this rule would not apply. But that isn't the desired meaning, right?
> If you intend to use a <code>label</code> element, place the

This way, there is no margin for interpretation about what to do with a label element.

CITguy replied:
good catch, will fix

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+    </div>
+    <rx-styleguide code-url="styleguide/forms/checkboxes.html"></rx-styleguide>
+
+    <h3 class="title ">Using a Character Counter</h3>
```
Extra space in the class.

CITguy replied:
good catch. this was a remnant of removing the `clear` class from the original markup (which is unnecessary)

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+      </ul>
+    </div>
+
+    <h3 class="title clear">Selects</h3>
```
I don't believe `clear` class is needed for this title, or any of the subsequent ones.

CITguy replied:
indeed, removed remaining `clear` classes. the only one that should remain is the one on line 2

megalord commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+        </li>
+        <li>
+          When label rules (seen above) are applied correctly, the label will also
+          get some styling when a radio or checkbox is disabled.
```
Too many "when"s in the same sentence.
> When label rules (seen above) are applied correctly to a radio or checkbox,
the label will also get some styling.

CITguy replied:
good call. will update with reference to toggle switches, too

megalord commented on src/rxFieldName/docs/rxFieldName.html:
```diff
@@ -28,41 +34,33 @@ <h2 class="title">Field Name States</h2>
+              <rx-field>
+                <rx-field-name id="eleFieldNameStateOptional">
+                  Optional Field
+                </rx-field_name>
```
`s/_/-/`

CITguy replied:
good catch. will fix

megalord commented on src/rxFieldName/rxFieldName.js:
```diff
@@ -1,11 +1,56 @@
+ * @param {Boolean=} [ng-required=false]
+ * Is field required? This will add/remove the required symbol to the left of the name.
+ */
+.directive('rxFieldName', function (rxNestedElement) {
```
`rxForm` now contains both `rxField` and `rxFieldContent`.  With this in mind, I don't see why `rxFieldName` gets its own component (i.e. directory and module).

CITguy replied:
@parlarjb, do you have any objection to moving rxFieldName into rxForm?

parlarjb replied:
No objection

CITguy replied:
TL;DR - I'll move it into `rxForm`

So, thinking about this some more, I realized why I split it into it's own component. I didn't want to clutter rxForm considering the recent refactoring to pull `rxOptionTable` *out* of rxForm.  If we had the ability to define sub-components in demo nav, display, etc., I would place the following components underneath `rxForm`

* rxCharacterCount
* rxCheckbox
* rxSelect
* rxMultiSelect
* rxRadio
* rxOptionTable
* rxToggleSwitch
* etc.

This way, all form control elements and directives would be housed within `rxForm`. However, our build process currently doesn't support this. However, considering the upcoming requirements to documentation, this probably wouldn't be a bad idea to look into.

For this PR, I'll move `rxFieldName` into `rxForm`.

parlarjb replied:
I'll say this: If an element has `rxForm` as a required controller (or has `rxForm` in a required controller "dependency chain"), then I'm fine with it living in rxForm. Ideally we'd make a little better use of Angular's module system, and at least put them in separate files, but our current build system wouldn't allow that.

CITguy replied:
Agreed. I'd love to be able to split each directive into its own source file. `rxForm.js` is getting lengthy with all of the documentation added for the directives.

megalord commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,194 @@
+layout), most of the new directives must be nested in a specific hierarchy.
+If you do not nest these elements properly, Angular will throw an error (this
+is by design). So, rule of thumb, aim for 0 console errors.  These directives
+must be nested in the proper order or angular will throw a console error.
```
This sentence is pretty much a duplicate of the one two lines above.

CITguy replied:
the intent was to be a segway to the list below, I'll make a tweak.

megalord commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,194 @@
+* [rxForm](/ngdocs/index.html#/api/rxForm.directive:rxForm)
+  * [rxFormSection](/ngdocs/index.html#/api/rxForm.directive:rxFormSection)
+    * [rxField](/ngdocs/index.html#/api/rxForm.directive:rxField)
+      * [rxFieldName](/ngdocs/index.html#/apirxFieldname.directive:rxFieldName)
```
This link is missing a `/` after `/api`.

CITguy replied:
good catch. will fix

megalord commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,194 @@
+can also use this element if you define a custom validator.
 
-# Services
+### Foocheck validator
```
Do we really need to include this?  It's basically an excerpt from the angular docs.

CITguy replied:
This was existing documentation. I'm merely bringing it forward.

megalord commented on src/rxForm/docs/rxForm.html:
```diff
@@ -1,117 +1,419 @@
-        <h3 class="title">Text inputs</h3>
-
-        <rx-form-item label="Plain textbox">
+<div class="" ng-controller="rxFormDemoCtrl">
```
Please remove the `class` attribute altogether.

CITguy replied:
good catch. will fix

megalord commented on src/rxForm/docs/rxForm.html:
```diff
@@ -1,117 +1,419 @@
+            <input rx-checkbox
+              id="chkNameRequired"
+              ng-model="simple.isNameRequired" />
+            <label for="chkNamerequired">
```
`s/r/R/`

CITguy replied:
good catch. will fix

megalord commented on src/rxForm/rxForm.js:
```diff
@@ -1,21 +1,516 @@
+ * ...
+ * </pre>
+ *
+ * @param {Boolean=} stacked
```
Labeling this param as a boolean is misleading.  It isn't really a boolean; it's *presence* is.  `stacked="false"`, `stacked="true"`, and `stacked` are no different.

CITguy replied:
Unfortunately, there's no concept of a boolean-by-presence attribute for ngdocs/jsdocs. This was the closest alternative I could find. Do you have any alternative suggestions?

CITguy replied:
@megalord, how would you feel about `@param {*=} stacked`?

megalord replied:
I tried several different syntaxes, but they all resulted in errors.  I guess this will have to do.

megalord commented on src/rxForm/rxForm.less:
```diff
@@ -307,3 +317,202 @@
+  rx-help-text,
+  rx-inline-error,
+  rx-input {
+    margin-bottom: (@rxForm-gutter / 2); // ~8px
```
This is the type of content that I doubt gets maintained, and could lead to confusion in the future.  How do you feel about removing it?

CITguy replied:
the comment, i presume?

CITguy replied:
will remove comment (and similar comments)

megalord commented on src/rxForm/rxForm.js:
```diff
@@ -1,21 +1,516 @@
+ * @restrict E
+ * @description
+ * Structural element directive used for layout of sub-elements.
+ * You may place an optional `.rx-form-actions` class on this element
```
I did a quick ctrl+f on this page, and this class isn't defined or used in the rest of the PR.  Is that by design?  If so, what purpose does the class serve?

CITguy replied:
remnant of old code. will remove

megalord commented on src/rxMisc/rxMisc.js:
```diff
@@ -115,38 +129,167 @@ angular.module('encore.ui.rxMisc', ['debounce', 'encore.ui.rxSessionStorage'])
+            opts.link = function () {};
+        }
 
+        return _.assign(defaults, opts);
```
I think there is a bug here.  `_.assign` mutates the destination object, so for subsequent function calls, the value of `defaults` is a mixture of it's previous value and the `opts` provided to the previous call.

Also, I think `_.defaults` better communicates your intention.

CITguy replied:
Ah, yes. I keep getting those two confused.

CITguy replied:
will change to `_.defaults`

megalord commented on src/styles/vars.less:
```diff
@@ -115,10 +115,21 @@
 @inputBackgroundError: @inputColorError;
 @inputBorderRadius: 0px;
-@inputBorderColor: #ccc;
+@inputBorderColor: #ccc; // use @rxForm-input-border-color
```
Was this comment intended for you to come back and fix before the PR is merged?

CITguy replied:
I was originally intending on adding deprecation comments, but I'll be leaving that discussion for another PR. will remove comment

parlarjb commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,340 @@
+      <ul class="list">
+        <li>Please use the <code>&lt;label for="formControlId"&gt;</code> format when defining your HTML labels.
+          <ul>
+            <li>When a <code>label</code> is clicked (wherever it is placed in the DOM), it should focus or activate its corresponding form control.</li>
```
How about mentioning that this is what the `for` attribute does? Or, at a minimum, change the wording. It currently sounds like making it focus/activate is something that the developer will have to figure out

CITguy replied:
good point. will update

CITguy commented on src/rxMisc/rxMisc.js:
```diff
@@ -115,38 +129,167 @@ angular.module('encore.ui.rxMisc', ['debounce', 'encore.ui.rxSessionStorage'])
+            opts.link = function () {};
+        }
 
+        return _.defaults(defaults, opts);
```
I realized I have the two arguments backward. This will be fixed in the next PR update.

Droogans commented:
Let me pull down this branch and see if I can't get it mergable. If everything that conflicts is just my changes, I should be able to get through it. Otherwise I'll grab @CITguy and see if we can't work through this by lunch.

glynnis commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,250 @@
+    <h3 class="title">Manual Form Saving</h3>
+    <div class="component-description">
+      <ul class="list">
+        <li>When an entity is being edited in place, rather than submitting an edit or create form, a notification with a save button is used.</li>
```
This is actually not the correct pattern. See https://projects.invisionapp.com/d/main#/projects/3750442

I'm not sure where @megalord and his team are with updating/changing this pattern on the dev side, but if it's still in progress I would prefer not documenting it until the correct pattern exists over documenting something that's incorrect.

megalord replied:
Ah, my bad on this.  I need to update the Encore UI docs to reflect the design change.  Though the diff shows an addition here, this content is no different than the existing documentation, which is also incorrect.  I'll put in a fix, but I don't think it should hold up this PR.

CITguy replied:
@megalord is correct, this looks like an addition, but it's the same content that is already in the [styleguide](http://rackerlabs.github.io/encore-ui/#/styleguide/forms) under "Manual Form Saving". Can the documentation update be submitted in a fast follow PR?

glynnis replied:
I'm good with a fast follow :+1:

CITguy commented on demo/assets/generated_demo.css:
```diff
@@ -105,3 +105,21 @@ table.component-attributes th,
+  padding: 0;
+  border: 0;
+  border-radius: 0;
+}
```
This is being added because it was missed in a previous PR that updated `demo/assets/demo.less`.

CITguy commented:
* `rxFieldName` moved within `rxForm`
* updated to reflect new syntax in page objects, midways, and exercise test files

Droogans commented:
LGTM. Overall, I'm ok with this as is. It is too large for a single pull request (as most refactor PRs tend to be), and there needs to be some more work on the rxForm page objects, but those don't need to be here in this PR. It can wait.

There are lots of new features to interact with. Some of the ones that stand out to me is all of the invalid/disabled states for form fields, as well as a better way to deal with groups of radio elements and groups of checkbox elements.

glynnis commented:
I'm with @Droogans -- LGTM on the design side :+1:, but wouldn't be surprised given the size of the PR if we'll need to make some tweaks and additions.

Thank you for this work @CITguy ! Very excited about the new form and how all this hard work is going to pay off :smile: :clap: :clap: :clap:

CITguy commented:
Updated to make use of `rx-help-text` within rxOptionTable to display the selected label text. This is to better differentiate the data value from the help text (selected label).

*Compare the position and style of "(Already saved data)" in the Before & After below.*

### Before
![screen shot 2015-06-26 at 10 58 43 am](https://cloud.githubusercontent.com/assets/545605/8384936/2e428638-1c0a-11e5-9904-a5427be0117a.png)

### After
![screen shot 2015-06-26 at 10 57 27 am](https://cloud.githubusercontent.com/assets/545605/8384942/3443565c-1c0a-11e5-8c47-8d48547ec0e6.png)

CITguy commented:
If all is well, please let me know when I should perform a squash for the final update.

Droogans commented:
Squash label's up! I say go for it.

parlarjb commented on demo/styleguide/forms/validation.html:
```diff
@@ -1,3 +1,71 @@
+        </rx-input>
+      </rx-field-content>
+
+      <!-- Uncomment if/when "invalid" styles available -->
```
Is there an issue filed for this?

CITguy replied:
Yes, https://github.com/rackerlabs/encore-ui/issues/1065

parlarjb commented on demo/styleguide/forms.html:
```diff
@@ -1,132 +1,250 @@
+          "ng-invalid" and/or "ng-dirty" classes (which are dynamically added by ngModel validation logic).
+        </li>
+        <li>Some components will style based on both classes, while others will only use <code>ng-invalid</code>.</li>
+        <li><strong>NOTE:</strong> Designing an EncoreUI specific design pattern for placement of form validation messages is on our roadmap.</li>
```
"Creating an EncoreUI-specific design pattern" sounds a little less awkward, but this is minor, I'll leave it up to you

CITguy replied:
Sounds good. will update.

parlarjb commented:
Could you add something for `[API Documentation](/ngdocs/index.html#/api/rxButton.directive:rxButton)` to the `component-template`?

CITguy commented:
@parlarjb, that may be something that will need to be addressed in #1063. Right now, the jsdoc paths are inconsistent, which would result in many links being broken.

parlarjb commented:
Ok, thanks

CITguy commented:
Squashed!

parlarjb commented on src/rxCheckbox/rxCheckbox.js:
```diff
@@ -1,4 +1,19 @@
+ *
+ * @example
+ * <pre>
+ * <input rx-checkbox />
```
How do you feel about adding more attributes here? I don't want someone to look at it and think that the `rx-checkbox` will handle the `ng-model` automatically.

CITguy replied:
ah, yes. I'll update.

parlarjb commented on src/rxCheckbox/rxCheckbox.spec.js:
```diff
@@ -0,0 +1,118 @@
+            });
+
+            it('should have "rxCheckbox" class', function () {
+                expect(parent.attr('class')).to.contain('rxCheckbox');
```
It's outside the scope of this PR, but what was your thinking on naming the class `rxCheckbox` instead of `rx-checkbox`? I can't recall if we spoke on that before.

CITguy replied:
We have element directives that are referenced as `rx-help-text` or `rx-field-name` (in spinal case), and IMO adding a dot to the beginning for a class name could be confusing (`.rx-field-name` vs `rx-field-name`).  So by using lowerCamelCase for class names, we avoid the similarity confusion (`.rxFieldName` vs `rx-field-name`).  But this is something that should be discussed around CSS standards (among other items).

Droogans replied:
I say follow Angular's [normalization naming methods](https://docs.angularjs.org/guide/directive#normalization).

CITguy replied:
The key thing is that the class names don't always come into play. 

### Attribute Directives
#### That wrap their target
For attribute directives that wrap their target element in custom markup (`rx-checkbox`, `rx-radio`, `rx-select`, etc.), we don't have a descriptive element for CSS selection.  In this case, I think it would be best to use the angular-normalized css name.

```LESS
.rxCheckbox { ... }
.rxSelect { ... }
.rxRadio { ... }
```

#### Others
For attribute directives that do not wrap their target element, use the attribute name selector.

```LESS
[rx-form] { ... }
```

### Element Directives
For element directives, use the element name.

```LESS
rx-field-name { ... }
rx-help-text { ... }
```

Droogans replied:
> IMO adding a dot to the beginning for a class name could be confusing (`.rx-field-name` vs `rx-field-name`)

That's as far as I had read the first time.

I see now that `.rxCheckbox` is signifying that they are wrapping the whole thing in a single class....at least that's what I'm understanding. `.rx-checkbox` would mean that it's just another class name inside of, say, a regular element directive `rx-checkbox`.

Is that closer to what's happening here?

CITguy replied:
I use `.rxCheckbox` because I needed to scope styles based on the target element's parent (CSS can't do this).  As such, a namespace needed to be created (`.rxCheckbox`) to properly scope the style definitions and not confuse the CSS selector as an element selector.

For other attribute and element directives, I can apply the style definitions downward using the attribute (`[rx-foobar]`) or element (`rx-foobar`) selector as the namespace.

CITguy replied:
To capture conversations around this topic, I've created #1079 for discussion.

parlarjb commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,204 @@
+      * [rxFieldContent](/ngdocs/index.html#/api/rxForm.directive:rxFieldContent)
+        * [rxInput](/ngdocs/index.html#/api/rxForm.directive:rxInput)
+          * [rxPrefix](/ngdocs/index.html#/api/rxForm.directive:rxPrefix)
+          * [rxSuffix](/ngdocs/index.html#/api/rxForm.directive:rxSuffix)
```
Can we added a bracketed number after each line to represent the "has-a" relationship here? i.e. `rxForm` can have `n` `rxFormSection` children, but each `rxField` can only have `1` `rxFieldName`,

```
rxForm
    rxFormSection (n)
        rxField (n)
            rxFieldName (1)
…
```

CITguy replied:
good idea. will do

parlarjb commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,204 @@
+* [rxRadio](/#/component/rxRadio)
+* [rxSelect](/#/component/rxSelect)
+* [rxToggleSwitch](/#/component/rxToggleSwitch)
+* [typeahead](/#/component/typeahead)
```
Do you know if `rxMultiSelect` works properly? If not, we should file an issue.

CITguy replied:
I've not tested `rxMultiSelect` with the `rxForm` revamp. However, `rxMultiSelect` uses `rxSelect` as a base, so I would imagine that it would be fine. That being said: yes, we should probably file an issue to **ensure** `rxMultiSelect` compatibility with `rxForm`.

CITguy replied:
Actually, yes! It does work.

![screen shot 2015-06-26 at 4 57 31 pm](https://cloud.githubusercontent.com/assets/545605/8388126/75d4e63e-1c24-11e5-917e-721f1ee72d30.png)

I'll add it as a compatible directive.

CITguy replied:
Added `rxMutliSelet` before `rxOptionTable`

parlarjb commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,204 @@
+## Before &amp; After
+The new markup may look a little wordy, but it is designed to provide flexibility
+(to a point) around content for advanced field inputs. To be explicit, the new
+directives were designed based on feedback around:
```
Please feel free to add something about `rxFormItem` being incredibly brittle and prone to breakage.

CITguy replied:
good call. will do

parlarjb commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,204 @@
 
-# Services
+## Required Fields
+When displaying a field should be required, please make use of the `ng-required`
```
`When displaying _that_ a field should be required`, or `When displaying a field _that_ should be required`

CITguy replied:
I like the second option. will update

CITguy commented:
Addressed feedback from @parlarjb and added examples around using `rxMultiSelect`.

![screen shot 2015-06-26 at 5 17 58 pm](https://cloud.githubusercontent.com/assets/545605/8388541/fbd839d0-1c28-11e5-89d6-e48e72bdf91a.png)

![screen shot 2015-06-26 at 5 18 08 pm](https://cloud.githubusercontent.com/assets/545605/8388544/0038634c-1c29-11e5-8bfc-a300d70addee.png)

parlarjb commented on src/rxForm/README.md:
```diff
@@ -1,45 +1,206 @@
+      * [rxFieldName](/ngdocs/index.html#/api/rxForm.directive:rxFieldName) (1)
+      * [rxFieldContent](/ngdocs/index.html#/api/rxForm.directive:rxFieldContent) (1)
+        * [rxInput](/ngdocs/index.html#/api/rxForm.directive:rxInput) (N)
+          * [rxPrefix](/ngdocs/index.html#/api/rxForm.directive:rxPrefix) (N)
```
What's the use case for having `N` `rxPrefix` or `rxSuffix` elements?

CITguy replied:
I'm not aware of any present use cases, but I didn't want our present designs to limit what we could potentially accomplish.  

#### Inspiration
I realize that it's not officially a supported framework pattern, but I wanted `rxPrefix` and `rxSuffix` to have similar functionality to how Bootstrap does ["Input Groups"](http://getbootstrap.com/components/#input-groups).  The only difference between EncoreUI and Bootstrap, in this regard, is styling.

parlarjb replied:
In Bootstrap, they say `We do not support multiple add-ons on a single side.` If we were to follow that, then these should be limited to 1, instead of `N`.

I also think it's more likely to confuse people with `N`, as it's counter to what people would expect to do.

If you do choose to keep `N`, then can you change the line above these examples to say "parens denote how many items _can_ be nested within its parent"

parlarjb replied:
A final note on this stuff: To be safe, we should probably say `0..N`, `0..1` for all of these, instead of just `N` and `1`. We don't want people to think that (for example) `rxPrefix` is required.

CITguy replied:
Ok, until we actually have a use case, I'll default to `1` on prefix/suffix. I'll update to using the `0..*` notation, as well.

parlarjb commented on src/rxForm/rxForm.js:
```diff
@@ -1,21 +1,568 @@
+ * </pre>
+ *
+ * @param {Boolean=} [ng-required=false]
+ * Is field required? This will add/remove the required symbol to the left of the name.
```
Can we change this to "Is _this_ field required?"

CITguy replied:
Sure. will do

parlarjb commented on src/rxForm/rxForm.js:
```diff
@@ -1,21 +1,568 @@
+ * @ngdoc directive
+ * @restrict E
+ * @description
+ * Sructural element directive used for layout of sub-elements.
```
"Structural"

CITguy replied:
Good catch. will fix

parlarjb commented on src/rxForm/rxForm.js:
```diff
@@ -1,21 +1,568 @@
+ *     <rx-field>
+ *       <rx-field-name>Salary:</rx-field-name>
+ *       <rx-field-content>
+ *         <rx-input>...</rx-input>
```
Could you please show an actual element inside of the `<rx-input>`, for clarity?

CITguy replied:
sure, will update

parlarjb commented on src/rxForm/rxForm.less:
```diff
@@ -307,3 +317,222 @@
+  box-sizing: border-box;
+  display: block;
+  margin: 0px;
+  padding: 0px;
```
These four lines are the same four as are defined above (for `[rx-form], `rx-form-section`, `rx-field-name`, etc.), do they need to be repeated?

CITguy replied:
ah, good catch. will remove duplication.

parlarjb commented on src/rxForm/rxForm.less:
```diff
@@ -307,3 +317,222 @@
+          }
+
+          input:not([type="checkbox"]):not([type="radio"]),
+          textarea {
```
Is there a styling reason that you put the `flex: 1 1 auto` into its own selector above, as opposed to putting it into this one?

CITguy replied:
ah, the previous definition was left over from some prototyping. will consolidate.

parlarjb commented:
LGTM

CITguy commented:
Wow, could Travis be any slower? :turtle:

megalord commented:
LGTM

halkon commented:
I got [̲̅$̲̅(̲̅ιοο̲̅)̲̅$̲̅] [̲̅$̲̅(̲̅ιοο̲̅)̲̅$̲̅] for this LGTM
